code/datums/cleaning.dm

```diff
@@ -0,0 +1,55 @@
+#define CMAG_CLEANTIME 50 //The cleaning time for cmagged objects is locked to this, for balance reasons
+
+
+//For handling standard click-to-clean items like soap and mops.
+
+//text_verb and text_description carry strings that, when pieced together by this proc, make the cleaning messages.
+//text_verb is the verb that'll be used for cleaning, i.e. "User begins to [mop] the floor."
+//text_description by default appends the name of the cleaning object to the cleaning message, i.e. "User begins to clean the floor[ with soap.]"
+//If text_description is ".", the sentence will end early (so you don't say "Lusty Xenomorph Maid begins to clean the floor with the Lusty Xenomorph Maid")
+
+/atom/proc/cleaning_act(mob/user, atom/cleaner, cleanspeed = 50, text_verb = "clean", text_description = " with [cleaner].")
+	var/is_cmagged = FALSE
+
+	if(user.client && (src in user.client.screen)) //You can't clean items you're wearing for technical reasons
+		to_chat(user, "<span class='notice'>You need to take that [name] off before cleaning it.</span>")
+		return FALSE
+
+	if(HAS_TRAIT(src, TRAIT_CMAGGED)) //So we don't need a cleaning_act for every cmaggable object
+		is_cmagged = TRUE
+		text_verb = "clean the ooze off"
+		cleanspeed = CMAG_CLEANTIME
+
+	user.visible_message("<span class='warning'>[user] begins to [text_verb] \the [name][text_description]</span>")
+	if(!do_after(user, cleanspeed, target = src) && src)
+		return FALSE
+
+	cleaner.post_clean(src, user)
+
+	if(!cleaner.can_clean())
+		return FALSE
```

Comment:
Shouldn't this go at/near the start of the proc? Things that can't clean probably shouldn't be able to get this far.

Reply:
`can_clean()` is where it is intentionally; Mops are an awkward case where you need to be able to "mop" (i.e. hit the do_after part of cleaning_act), but fail to actually clean decals if your mop doesn't have cleaning reagents inside it. This is because mops can carry arbitrary reagents and the code is set up to allow you to mop with them even if you wouldn't be able to clean the floor, so you can do gimmicky stuff like mop floors with fluorosulfuric acid.

Mop is currently the only cleaning item that makes use of `can_clean` in this fashion: In all other instances, cleaning items which call clean_act are able to unconditionally clean decals at all times.